File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    
    const dx = [0,0,-1,1]
    const dy = [1,-1,0,0]
    
    const visited = Array.from({length:maps.length}, () => Array.from({length:maps[0].length}, () => false))
    
    const queue = [[0,0,1]]
    visited[0][0] = true
    
    while(queue.length > 0){
        const [x,y,dist] = queue.shift()
        
        if(x === maps.length-1 && y === maps[0].length-1){
            return dist
        }
        
        for(let i = 0; i < 4; i++){
            const nx = dx[i] + x
            const ny = dy[i] + y
            
            if(0 <= nx && nx < maps.length && 0 <= ny && ny < maps[0].length){
                if(!visited[nx][ny] && maps[nx][ny]){
                    visited[nx][ny] = true
                    queue.push([nx, ny, dist + 1])
                    
                }
            }
        }
    }
    return -1;
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    answer = -1
    m = len(maps)       # 세로
    n = len(maps[0])    # 가로
    
    visited = [[False] * n for _ in range(m)]
    queue = deque()
    queue.append([0,0,1])
    
    dx = [0,0,1,-1]
    dy = [1,-1,0,0]
    
    while queue :
        x,y,c = queue.popleft()
        if x == m - 1 and y == n - 1 :
            return c
        
        for i in range(4) :
            nx = x + dx[i]
            ny = y + dy[i]

            if nx >= 0 and nx < m and ny >= 0 and ny < n :
                if visited[nx][ny] == False and maps[nx][ny] == 1:
                    visited[nx][ny] = True
                    queue.append([nx,ny,c + 1])
                
    return answer